compiler/optimizations/gpuTransforms.cpp

```diff
@@ -136,6 +136,61 @@ static bool isDegenerateOuterRef(Symbol* sym, CForLoop* loop) {
   return true;
 }
 
+static bool callsInBodyAreGpuizableHelp(BlockStmt* blk,
```

Comment:
Any particular reason you lifted this out of the class? I suppose the argument is it makes it clear you're not mutating anything (other than the okFns set being passed in by reference).

Actually wait...

Looking at this, shouldn't visitedFns be passed by reference too?

Reply:
> Any particular reason you lifted this out of the class? I suppose the argument is it makes it clear you're not mutating anything (other than the okFns set being passed in by reference).

Because.... I don't know :) As a helper to facilitate recursion, it didn't feel like it is part of the interface of the class. Public or private. ie, why did we have other static functions in this file, they could also be private functions of the class.

But, I am not the person to champion better OOP practices. I can make it a private function if you think that's a better practice.

> Looking at this, shouldn't visitedFns be passed by reference too?

Absolutely, it should. It wasn't before this PR and we are already paying for it in compilation time.

FWIW, I think we should measure the time it takes for us to do gpuization. I created https://github.com/Cray/chapel-private/issues/3600 to record that and some other performance testing ideas.

Reply:
> Because.... I don't know :) As a helper to facilitate recursion, it didn't feel like it is part of the interface of the class. Public or private. ie, why did we have other static functions in this file, they could also be private functions of the class.

My sense is since this is a utility only meant to be used by that class it's better cohesion to have it as part of the class.  I like having the things that go together packaged together.  This class isn't being declared in a public header so I'm ok with it having a lot of implementation-specific private members.

As far as why aren't the other utiltities packaged in classes: I guess I'd say some of them really are things that aren't so heavily tied to a specific class and some of it's just the way it evolved.

I think the argument "it's better not to make it a non-const non-static member function unless its mutating data" is a good one (the more I can guess the side effects of a function based of its signature the better it is).  You can accomplish the same thing by making it a static member method.  So maybe the right thing to do is make it a static method of the class?

> But, I am not the person to champion better OOP practices. I can make it a private function if you think that's a better practice.

Well, I don't know if I'm that person either :-).  I was asking my question about "why you did it that way" earnestly, I suspect there's multiple "best practice" approaches for this sort of "how to package/organize" functions so I was curious if there was a reason.

Anyway, I'd say if there's not some reason to do otherwise, my preference would be that it was a private static method.

>> Looking at this, shouldn't visitedFns be passed by reference too?
> Absolutely, it should. It wasn't before this PR and we are already paying for it in compilation time.

Woohoo! I love it when code review actually finds a technical issue.

> FWIW, I think we should measure the time it takes for us to do gpuization. I created https://github.com/Cray/chapel-private/issues/3600 to record that and some other performance testing ideas.

Yeah that would be nice. I think I recall saying that `isDefinedInTheLoop` might be leading to an O(n^2) kind of behavior (but it wasn't clear to me that n would get sufficiently large to matter). It'd be great to get performance data.